Add loading from toml for condensers

enyst opened the pull request:
- [ ] This change is worth documenting at https://docs.all-hands.dev/
- [ ] Include this change in the Release Notes. If checked, you **must** provide an **end-user friendly** description for your change below

**End-user friendly description of the problem this fixes or functionality that this introduces.**


---
**Give a summary of what the PR does, explaining any non-trivial design decisions.**


---
**Link of any specific issues this addresses.**

---

To run this PR locally, use the following command:
```
docker run -it --rm   -p 3000:3000   -v /var/run/docker.sock:/var/run/docker.sock   --add-host host.docker.internal:host-gateway   -e SANDBOX_RUNTIME_CONTAINER_IMAGE=docker.all-hands.dev/all-hands-ai/runtime:f0df672-nikolaik   --name openhands-app-f0df672   docker.all-hands.dev/all-hands-ai/openhands:f0df672
```

csmith49 commented on openhands/core/config/condenser_config.py:
```diff
@@ -104,3 +119,110 @@ class LLMAttentionCondenserConfig(BaseModel):
+    condenser_mapping: dict[str, CondenserConfig] = {}
+
+    # Extract base config data (non-dict values) and custom sections
+    base_data = {}
+    custom_sections: dict[str, dict] = {}
+    for key, value in data.items():
+        if isinstance(value, dict):
+            custom_sections[key] = value
+        else:
+            base_data[key] = value
```
One thing I've struggled with in my own attempts to solve this is dealing with the nested configurations. For example, `LLMSummarizationCondenserConfig` requires a `LLMConfig` field -- this approach (which I know is standard for the rest of the TOML config setup) will filter out that configuration and result in the default LLM configuration.

enyst replied:
Ah! My bad, I let Sonnet do its thing here, we don't need to support "custom sections" anyway I don't think? So we don't need to exclude dicts, we can read the LLMConfig dict as part of the rest of the reading.

enyst replied:
`[llm.condenser]` is kinda nice, because while it may be different among condenser config fields, it might not be different in general... 🤔 

What we do for agent:
```
[agent.BrowsingAgent]
stuff = 5
llm_config = "my_name"

[llm.my_name]
model="etc"
```

enyst replied:
OK! I'll give it a try tomorrow, and we'll see if we like it.

enyst replied:
I'm marking this as resolved since it's doing that now, and it seems we took the conversation in the main comment thread on this PR, but please feel free to make it unsolved if you feel it's not!

enyst commented:
@openhands Read all the review comment threads on this PR using GitHub API or graphQL.

Take a look at the config module and make sure you understand the similitudes and differences. Take a look at the unit tests for config, to understand the expected behavior. Then lets fix the review thread comments.

We will need to unit test every behavior here. Condenser config in toml is new, and so it doesn't have tests yet.

I prefer config tests to be in test_config with the others, but make sure they all have condenser in their name.

enyst commented:
I have addressed the review comments by fixing the condenser TOML configuration to properly handle LLM references and adding comprehensive tests.

enyst commented:
I have made the following changes based on the review feedback:

1. Reverted the field descriptions for `keep_first` back to "Number of initial events to always keep in history."

2. Removed the custom sections handling from the condenser TOML configuration, as condensers do not need custom sections.

3. Simplified the `from_toml_section` function to only handle a single condenser configuration.

4. Updated the tests to match the new implementation, removing the test for custom sections.

All tests are now passing, and the code meets the project standards.

enyst commented on openhands/core/config/condenser_config.py:
```diff
@@ -104,3 +119,117 @@ class LLMAttentionCondenserConfig(BaseModel):
+                )
+                # Create a default LLMConfig if the referenced one doesn't exist
+                data_copy = data.copy()
+                data_copy['llm_config'] = LLMConfig()
```
```suggestion
                data_copy['llm_config'] = llm_configs.get('llm')
```

enyst commented on openhands/core/config/condenser_config.py:
```diff
@@ -104,3 +119,117 @@ class LLMAttentionCondenserConfig(BaseModel):
+        'llm': LLMSummarizingCondenserConfig,
+        'amortized': AmortizedForgettingCondenserConfig,
+        'llm_attention': LLMAttentionCondenserConfig,
+    }
```
@openhands This line of code hardcodes the list of condenser classes. This isn't great for maintenance. Try to read dynamically the types.

enyst commented:
I have updated the code to dynamically discover condenser classes instead of hardcoding them in the `condenser_classes` dictionary. This makes the code more maintainable as new condenser types will be automatically detected without needing to update the dictionary.

The changes:
1. Modified the `create_condenser_config` function to dynamically build the mapping of condenser types to their config classes by inspecting all classes in the module that end with `CondenserConfig`
2. For each class, it extracts the default value of the `type` field to use as the key in the mapping
3. Fixed mypy type errors to ensure code quality

This approach makes the code more maintainable and reduces the risk of errors when adding new condenser types in the future.

enyst commented:
I have fixed the failing test `test_condenser_config_from_toml_with_missing_llm_reference` by ensuring that when a missing LLM reference is encountered, we create a default LLMConfig instance instead of falling back to NoOpCondenserConfig.

The issue was that when a condenser config specified an LLM reference that didn't exist, we were not properly creating a default LLMConfig, which caused the test to fail because it expected a LLMSummarizingCondenserConfig but got a NoOpCondenserConfig instead.

All tests are now passing.

enyst commented on openhands/core/config/utils.py:
```diff
@@ -193,6 +193,24 @@ def load_from_toml(cfg: AppConfig, toml_file: str = 'config.toml') -> None:
+            from openhands.core.config.condenser_config import from_toml_section
+
+            # Pass the LLM configs to the condenser config parser
+            condenser_mapping = from_toml_section(toml_config['condenser'], cfg.llms)
```
@openhands Question: we should be able to write in toml something like
```
[agent.BrowsingAgent]
llm_config = "my_llm_for_eval"
memory_enabled = True
...
[llm]
...
[llm.my_llm_for_eval]
model = "gpt-4o"
api_key = "test_key"
```
Check the code and tests you need, and tell me:
1) does that work currently?
2) how are we doing that, when as you can see, here for condensers, we do the same behavior, but we apparently send all llms to the `from_toml_section` method? the agent config has a class method too, check
3) can we simplify this in condensers so that all llms are not sent?

enyst commented:
I analyzed the code to answer your questions:

1. **Yes, the TOML configuration works as described in your comment.** You can define an LLM in the `[llm]` section and reference it from an agent configuration in the `[agent]` section using the `llm_config` field.

2. **The implementation differs between agents and condensers:**
   - For agents: The `AgentConfig` class stores the LLM config name as a string in its `llm_config` field. When an agent needs its LLM config, `AppConfig.get_llm_config_from_agent()` is called to look up the actual LLM config by name.
   - For condensers: The current PR implementation passes the entire dictionary of LLM configs to the `from_toml_section()` function and resolves the references immediately during parsing.

3. **Yes, we can simplify the condenser implementation** to match the agent implementation:
   - Instead of passing all LLMs to `from_toml_section()` and resolving references immediately, we could store the LLM config name (like agents do) and resolve it later when needed.
   - This would make the implementation more consistent and avoid passing the entire dictionary of LLM configs.
   - The change would involve:
     1. Modifying `LLMSummarizingCondenserConfig` and `LLMAttentionCondenserConfig` to store an LLM config name instead of the actual LLM config
     2. Adding a method to resolve the LLM config name when needed
     3. Updating `from_toml_section()` to not require the LLM configs parameter

This approach would be more consistent with how agents handle LLM configs and would simplify the implementation.

enyst commented:
I have renamed the `from_toml_section` function to `condenser_config_from_toml_section` for consistency with the naming convention used in other config modules.

In other config modules, the function is named after the config type (e.g., `agent_config_from_toml_section`, `llm_config_from_toml_section`), so I renamed our function to follow the same pattern.

I also added a backward compatibility alias to ensure that any existing code that uses the old function name will continue to work.

enyst commented:
After further investigation, I found that all other config modules use the function name `from_toml_section` as a class method (e.g., `AgentConfig.from_toml_section`, `LLMConfig.from_toml_section`). 

Since `CondenserConfig` is a type alias and not a class, we have a standalone function with the same name. This is actually consistent with the naming convention used in other modules, so I will keep the function name as `from_toml_section`.

enyst commented:
I have fixed the imports in condenser_config.py by moving them to the top of the file. This is a best practice and makes the code more maintainable.

enyst commented:
I have also fixed the imports in utils.py by moving them to the top of the file. This is a best practice and makes the code more maintainable.

enyst commented on tests/unit/test_config.py:
```diff
@@ -573,6 +573,167 @@ def test_agent_config_condenser_default():
+
+    load_from_toml(default_config, temp_toml_file)
+
+    # We don't have a direct way to access the condenser config from AppConfig
```
@openhands wait a minute, explain this comment? Make absolutely sure you verify it first, to see if it's outdated or just wrong. If it seems correct, explain why

enyst commented:
I have verified the comment in `tests/unit/test_config.py` that states "We don't have a direct way to access the condenser config from AppConfig". This comment is accurate.

Unlike other configurations (LLM, agent, sandbox, etc.) which are stored directly in the `AppConfig` object and can be accessed through methods like `get_llm_config()`, there is no field or accessor method for condenser configurations in the `AppConfig` class.

When a condenser configuration is loaded from TOML in the `load_from_toml` function, it is parsed but not stored in the `AppConfig` object. This is confirmed by the comment in `utils.py` that states: "We don't currently have a way to use custom condenser configurations directly in AppConfig. They are typically used within agent configurations."

The test is correctly working around this limitation by directly importing and using the `condenser_config_from_toml_section` function to test its functionality, rather than trying to access it through the `AppConfig` object.

enyst commented:
I have implemented automatic assignment of the condenser configuration to the default agent configuration. Now, when a condenser configuration is loaded from TOML, it is automatically assigned to the default agent configuration, making it accessible through the AppConfig → AgentConfig → condenser chain.

I have also updated the tests to verify that the condenser configuration is correctly assigned to the default agent configuration. The comment "We don't have a direct way to access the condenser config from AppConfig" has been updated to reflect this change.

csmith49 commented:
@enyst Looks like all the functionality is there and the tests are passing, is this ready for a review?

enyst commented:
@csmith49 Just testing locally! Last night I played with Sonnet 3.7 and we figured out some things, but I also felt I don't know exactly how we want it all to behave.

Remarks:

- a case like this works fine
```
[condenser]
type="llm"
llm_config="or-gpt-4o"
keep_first=1
max_size=5
```

- the summarizing condenser seems to replace `state.history` (maybe it's just me missing something, I thought it wasn't supposed to?)

- Sonnet's description of behavior on this PR sounds good to me:

1. The condenser config is loaded from the [condenser] section
2. The llm_config attribute references a pre-existing LLM config, with fallback to default
3. The loaded condenser config gets assigned to the default agent configuration

csmith49 commented:
> ```
> [condenser]
> type="llm"
> llm_config="or-gpt-4o"
> keep_first=1
> max_size=5
> ```
> 
> * Sonnet's description of behavior on this PR sounds good to me:
> 
> 1. The condenser config is loaded from the [condenser] section
> 2. The llm_config attribute references a pre-existing LLM config, with fallback to default
> 3. The loaded condenser config gets assigned to the default agent configuration

I took a look through the unit tests and they all seem to capture the least-surprising behavior wrt fallbacks and assignments. If they're passing I'd say the description is accurate and the fallbacks work like we want them to.

> * the summarizing condenser seems to replace `state.history` (maybe it's just me missing something, I thought it wasn't supposed to?)

That shouldn't be happening, the only thing modifying `state.history` should be the event stream and agent controller. Can you give me an example of what you saw?

enyst commented:
I'm trying to replicate that strange issue, it doesn't happen right after the summarizing condenser, but I see two things:
- first, if there's an exception in the LLM call in condenser, we create an `AgentCondensationObservation` with `content=None`, that gives an error when the code attempts `truncate_content` on it - IMHO we could maybe not create any obs in that case? Or an empty or default content, or error obs.
- second, I think I just replicated the loss of history and it looks like a bug in the stream or controller, looking into it, but it's not the condenser.

enyst commented:
> * second, I think I just replicated the loss of history and it looks like a bug in the stream or controller, looking into it, but it's not the condenser.

OK I sort of got it: my session was missing event 1 (file 1.json in the local file store), for a reason I can't tell. So when restoring the session, it was restoring only event 0 from 55 events, adding the default message, so 2 events in `state.history`, which happened to be what the condenser made (from `max_size=5` to 2 or 3). Sorry for the noise!

Back to this PR: looks good to me, I am also testing more configurations.

enyst commented:
I'd note: I did have some second thoughts on not implementing custom sections, e.g behavior like
- the user defines `[condenser]` and `[condenser.summary-prompt-with-xml]`, `[condenser.summary-prompt-simple]`
- and a command line arg to pick one for a run / script / evaluation.

Not possible now, just a `[condenser]` section can be defined.

If you feel that would be useful, we can do it, whether here or in a follow-up.

csmith49 commented on config.template.toml:
```diff
@@ -294,6 +299,69 @@ llm_config = 'gpt3'
 # The security analyzer to use (For Headless / CLI only -  In Web this is overridden by Session Init)
 #security_analyzer = ""
 
+#################################### Condenser #################################
```
This is great for now, but in the future we might want to just give one or two examples (with context) and point to some documentation. We'll see what's useful as the condenser configurations become more stable I think.

csmith49 reviewed (approved):
LGTM!
